# Clase 1 

In [1]:
grosor_billete = 0.1 * 0.001
altura_obelisco = 67.5
num_billetes = 1
dia = 1

while grosor_billete * num_billetes <= altura_obelisco:
    dia = dia + 1
    num_billetes = num_billetes * 2

print('Cantidad de días: ', dia)
print('Cantidad de billetes: ', num_billetes)
print('Altura final: ', num_billetes * grosor_billete)


Cantidad de días:  21
Cantidad de billetes:  1048576
Altura final:  104.8576


In [6]:
nombre = input("¿Cuál es tu nombre?")
print("Hola", nombre)

¿Cuál es tu nombre?noeli
Hola noeli


#### Ejercicio 1.5: La pelota que rebota
Una pelota de goma es arrojada desde una altura de 100 metros y cada vez que toca el piso salta 3/5 de la altura desde la que cayó. Escribí un programa rebotes.py que imprima una tabla mostrando las alturas que alcanza en cada uno de sus primeros diez rebotes.

In [2]:
altura = 100
contador = 0

while contador <= 10:
    print(contador, round(altura, 4))
    altura = altura * 0.6
    contador = contador + 1

0 100
1 60.0
2 36.0
3 21.6
4 12.96
5 7.776
6 4.6656
7 2.7994
8 1.6796
9 1.0078
10 0.6047


#### Ejercicio 1.6: Saludos
Escribí un programa llamado saludo.py que pregunte el nombre de le usuarie, imprima un saludo (por ejemplo, "Hola, Juana") y termine.

#### Ejercicio 1.7: La hipoteca de David
David solicitó un crédito a 30 años para comprar una vivienda, con una tasa fija nominal anual del 5%. Pidió $500000 al banco y acordó un pago mensual fijo de $2684,11.
El siguiente es un programa que calcula el monto total que pagará David a lo largo de los años:

In [2]:
saldo = 500000.0
tasa = 0.05
pago_mensual = 2684.11
total_pagado = 0.0

while saldo > 0:
    saldo = saldo * (1+tasa/12) - pago_mensual
    total_pagado = total_pagado + pago_mensual

print('Total pagado', round(total_pagado, 2))

Total pagado 966279.6


#### Ejercicio 1.8: Adelantos
Supongamos que David adelanta pagos extra de $1000/mes durante los primeros 12 meses de la hipoteca.

Modificá el programa para incorporar estos pagos extra y que imprima el monto total pagado junto con la cantidad de meses requeridos.

Cuando lo corras, este nuevo programa debería dar un pago total de 929965.62 en 342 meses.

Aclaración: aunque puede parecer sencillo, este ejercicio requiere que agregues una variable mes y que prestes bastante atención a cuándo la incrementás, con qué valor entra al ciclo y con qué valor sale del ciclo. Una posiblidad es inicializar mes en 0 y otra es inicializarla en 1. En el primer caso es problable que la variable salga del ciclo contando la cantidad de pagos que se hicieron, en el segundo, ¡es probable que salga contando la cantidad de pagos más uno!

In [20]:
saldo = 500000.0
tasa = 0.05
pago_mensual = 2684.11
total_pagado = 0.0
mes = 0

while saldo > 0:
    if mes <= 12:
        saldo = saldo * (1+tasa/12) - (pago_mensual + 1000)
    else:
        saldo = saldo * (1+tasa/12) - pago_mensual 
        
    total_pagado = total_pagado + pago_mensual
    mes = mes + 1

print('Total pagado', round(total_pagado, 2), 'en', mes, 'meses')

Total pagado 912597.4 en 340 meses


#### Ejercicio 1.9: Calculadora de adelantos
¿Cuánto pagaría David si agrega $1000 por mes durante cuatro años, comenzando en el sexto año de la hipoteca (es decir, luego de 5 años)?

Modificá tu programa de forma que la información sobre pagos extras sea incorporada de manera versatil. Agregá las siguientes variables antes del ciclo, para definir el comienzo, fin y monto de los pagos extras:

In [24]:
saldo = 500000.0
tasa = 0.05
pago_mensual = 2684.11
total_pagado = 0.0
pago_extra_mes_comienzo = 61
pago_extra_mes_fin = 108
pago_extra = 1000
mes = 0

while saldo > 0:
    if mes >= pago_extra_mes_comienzo and  mes <= pago_extra_mes_fin:
        saldo = saldo * (1+tasa/12) - (pago_mensual + pago_extra)
        total_pagado = total_pagado + pago_mensual + pago_extra
    else:
        saldo = saldo * (1+tasa/12) - pago_mensual
        total_pagado = total_pagado + pago_mensual
        
    
    mes = mes + 1

print('Total pagado', round(total_pagado, 2), 'en', mes, 'meses')

Total pagado 880074.1 en 310 meses


#### Ejercicio 1.10: Tablas
Modicá tu programa para que imprima una tabla mostrando el mes, el total pagado hasta el momento y el saldo restante.

Debería dar:
Total pagado:  880074.1
Meses:  310

In [29]:
saldo = 500000.0
tasa = 0.05
pago_mensual = 2684.11
total_pagado = 0.0
pago_extra_mes_comienzo = 61
pago_extra_mes_fin = 108
pago_extra = 1000
mes = 0

while saldo > 0:
    if mes >= pago_extra_mes_comienzo and  mes <= pago_extra_mes_fin:
        saldo = saldo * (1+tasa/12) - (pago_mensual + pago_extra)
        total_pagado = total_pagado + pago_mensual + pago_extra
    else:
        saldo = saldo * (1+tasa/12) - pago_mensual
        total_pagado = total_pagado + pago_mensual
    if saldo <= 0:
        saldo = 0
        
    mes = mes + 1
    print(mes, round(total_pagado, 2), round(saldo, 2))

print('Total pagado:', round(total_pagado, 2))
print('Meses:', mes)


1 2684.11 499399.22
2 5368.22 498795.94
3 8052.33 498190.15
4 10736.44 497581.83
5 13420.55 496970.98
6 16104.66 496357.58
7 18788.77 495741.63
8 21472.88 495123.11
9 24156.99 494502.01
10 26841.1 493878.33
11 29525.21 493252.04
12 32209.32 492623.15
13 34893.43 491991.64
14 37577.54 491357.49
15 40261.65 490720.7
16 42945.76 490081.26
17 45629.87 489439.16
18 48313.98 488794.38
19 50998.09 488146.91
20 53682.2 487496.75
21 56366.31 486843.87
22 59050.42 486188.28
23 61734.53 485529.96
24 64418.64 484868.89
25 67102.75 484205.06
26 69786.86 483538.47
27 72470.97 482869.11
28 75155.08 482196.95
29 77839.19 481522.0
30 80523.3 480844.23
31 83207.41 480163.64
32 85891.52 479480.21
33 88575.63 478793.93
34 91259.74 478104.8
35 93943.85 477412.79
36 96627.96 476717.9
37 99312.07 476020.11
38 101996.18 475319.42
39 104680.29 474615.81
40 107364.4 473909.27
41 110048.51 473199.78
42 112732.62 472487.33
43 115416.73 471771.92
44 118100.84 471053.53
45 120784.95 470332.14
46 123469.06 469607.75

#### Ejercicio 1.11: Bonus
Ya que estamos, corregí el código anterior de forma que el pago del último mes se ajuste a lo adeudado.

In [1]:
saldo = 500000.0
tasa = 0.05
pago_mensual = 2684.11
total_pagado = 0.0
pago_extra_mes_comienzo = 61
pago_extra_mes_fin = 108
pago_extra = 1000
mes = 0

while saldo > pago_mensual:
    if mes >= pago_extra_mes_comienzo and  mes <= pago_extra_mes_fin:
        saldo = saldo * (1+tasa/12) - (pago_mensual + pago_extra)
        total_pagado = total_pagado + pago_mensual + pago_extra
    else:
        saldo = saldo * (1+tasa/12) - pago_mensual
        total_pagado = total_pagado + pago_mensual
    mes = mes + 1
    print(mes, round(total_pagado, 2), round(saldo, 2))
    

if  saldo <= pago_mensual:
    pago_mensual = saldo
    saldo = saldo - pago_mensual
    total_pagado = total_pagado + pago_mensual
    print(mes, round(total_pagado, 2), round(saldo, 2))

print(f"Total pagado: ${round(total_pagado, 2)} en {mes} meses")


1 2684.11 499399.22
2 5368.22 498795.94
3 8052.33 498190.15
4 10736.44 497581.83
5 13420.55 496970.98
6 16104.66 496357.58
7 18788.77 495741.63
8 21472.88 495123.11
9 24156.99 494502.01
10 26841.1 493878.33
11 29525.21 493252.04
12 32209.32 492623.15
13 34893.43 491991.64
14 37577.54 491357.49
15 40261.65 490720.7
16 42945.76 490081.26
17 45629.87 489439.16
18 48313.98 488794.38
19 50998.09 488146.91
20 53682.2 487496.75
21 56366.31 486843.87
22 59050.42 486188.28
23 61734.53 485529.96
24 64418.64 484868.89
25 67102.75 484205.06
26 69786.86 483538.47
27 72470.97 482869.11
28 75155.08 482196.95
29 77839.19 481522.0
30 80523.3 480844.23
31 83207.41 480163.64
32 85891.52 479480.21
33 88575.63 478793.93
34 91259.74 478104.8
35 93943.85 477412.79
36 96627.96 476717.9
37 99312.07 476020.11
38 101996.18 475319.42
39 104680.29 474615.81
40 107364.4 473909.27
41 110048.51 473199.78
42 112732.62 472487.33
43 115416.73 471771.92
44 118100.84 471053.53
45 120784.95 470332.14
46 123469.06 469607.75

#### Ejercicio 1.12: Un misterio
Explicar el siguiente comportamiento:

In [48]:
bool("False")

True

Estamos pasando el string "False", no el valor False. Por ende, al ser un valor, es True.

#### Ejercicio 1.13: El volúmen de una esfera
En tu directorio de trabajo, escribí un programa llamado esfera.py que le pida al usuario que ingrese por teclado el radio r de una esfera y calcule e imprima el volumen de la misma. Sugerencia: recordar que el volúmen de una esfera es 4/3 πr^3.

Finalmente, ejecutá el programa desde la línea de comandos para responder ¿cuál es el volumen de una esfera de radio 6? Debería darte 904.7786842338603.

In [15]:
from math import pi

r = float(input("Ingrese el radio de una esfera: "))
volumen = 4/3 * pi * r ** 3

print("El volumen de la esfera es:", volumen)

Ingrese el radio de una esfera: 6
El volumen de la esfera es: 904.7786842338603


## 1.4 Cadenas

In [5]:
#slice
a = 'Hello world'
d = a[:5]     # 'Hello'
e = a[6:]     # 'world'
f = a[3:8]    # 'lo wo'
g = a[-5:]    # 'world'

In [13]:
a[-5:]

'world'

In [23]:
 frutas = 'Manzana,Naranja,Mandarina,Banana,Kiwi'

In [16]:
frutas[0]

'M'

In [17]:
frutas[-1]

'i'

In [18]:
frutas[0] = 'm' # Los strings son solo de lectura

TypeError: 'str' object does not support item assignment

In [24]:
frutas = frutas + ',Pera'
frutas

'Manzana,Naranja,Mandarina,Banana,Kiwi,Pera'

In [26]:
frutas = 'Melón,'+frutas
frutas

'Melón,Melón,Manzana,Naranja,Mandarina,Banana,Kiwi,Pera'

In [27]:
'Naranja' in frutas

True

In [28]:
'nana' in frutas

True

In [29]:
'Lima' in frutas

False

#### Ejercicio 1.17: Iteración sobre cadenas

In [30]:
cadena = 'Ejemplo con for'

for c in cadena:
    print('catacter:', c)

catacter: E
catacter: j
catacter: e
catacter: m
catacter: p
catacter: l
catacter: o
catacter:  
catacter: c
catacter: o
catacter: n
catacter:  
catacter: f
catacter: o
catacter: r


Modificá el código anterior de manera que dentro del ciclo el programa cuente cuántas letras "o" hay en la cadena.

In [39]:
cadena = 'Ejemplo con for'
contador = 0

for c in cadena:
    if 'o' in c:
        contador = contador + 1

print('Hay', contador, 'letras "o"')  

Hay 3 letras "o"


In [ ]:
#### Ejercicio 1.18: Geringoso rústico

In [108]:
#cadena = 'Geringoso'
cadena = 'Geringoso'
capadepenapa = ''
vocales = "aeiou"

for c in cadena: 
    for vocal in vocales:
        if c == vocal:
            c = c.replace(vocal, vocal + 'p' + vocal)
    capadepenapa = capadepenapa + c
    print(c, end='')
  

Geperipingoposopo

In [109]:
frutas.lower()

'melón,melón,manzana,naranja,mandarina,banana,kiwi,pera'

In [110]:
frutas

'Melón,Melón,Manzana,Naranja,Mandarina,Banana,Kiwi,Pera'

In [111]:
frutas.find('Mandarina')

28

In [112]:
frutas[13:17]

'anza'

In [114]:
nombre = '   Naranja   \n'
nombre

'   Naranja   \n'

In [116]:
nombre = nombre.strip()
nombre

'Naranja'

#### Ejercicio 1.20: f-strings
Modificá el programa hipoteca.py del Ejercicio 1.11 de la sección anterior para que escriba su salida usando f-strings. Tratá de hacer que la salida quede bien alineada.

#### Ejercicio 1.29: Traductor (rústico) al lenguaje inclusivo
Queremos hacer un traductor que cambie las palabras masculinas de una frase por su versión neutra. Como primera aproximación, completá el siguiente código para reemplazar todas las letras 'o' que figuren en el último o anteúltimo caracter de cada palabra por una 'e'. Por ejemplo 'todos somos programadores' pasaría a ser 'todes somes programdores'. Guardá tu código en el archivo inclusive.py

In [55]:
frase = 'todes somos programadores y cientifico'
palabras = frase.split()
index1 = -1
index2 = -2
frase_t = ''
    
for palabra in palabras:
    if len(palabra) >= 2:
        if palabra[index1] == 'o':
            palabra = palabra[:index1] + 'e' 
            pass
        elif palabra[index2] == 'o' :
            palabra = palabra[:index2] + 'e' + palabra[index2+1:]  
        else:
            pass
    frase_t += palabra + ' '
    
print(frase_t)


['todes', 'somos', 'programadores', 'y', 'cientifico']
todes somes programadores y cientifice 
